### PR TITLE
Add config & arg for @paginate to limit use of count

### LINKED ISSUE #503: Add option to limit "count" in @paginate directive
When using `@paginate` directive the end user is able to request an unlimited amount of records from the GraphQL API. This can result in timeouts and high DB load.

For huge tables it would be good to be able to restrict the `count`-argument to some maximum value.

**Proposed solution**

- Add global lighthouse config with default maximum (`null` for unlimited, default to remain backwards compatibility)
- Add `maxCount` argument to `@paginate`-directive (overrides global config)

***Behavior***
- If `count` is set to to more than maximum throw Error 
Alternatively: Reduce `count` to maximum allowed if higher value is provided by the user.
What do you prefer?

Is this something you would accept an PR for?

### PR DESCRIPTION
**Pull request by jbbr**
**Related Issue/Intent**

Resolves #503

**Changes**

- Adds global config `lighthouse.paginate_max_count`
  defaults to `null`. `null` means "unlimited" (decided against `false` to be able to typehint with `int`)

- Adds `maxCount`-arg to `@paginate`-directive which overwrites global config

**Todo**

- [x] maxCount for `@paginate`
- [x] maxCount for `@hasMany`
- [x] Added or updated tests
- [x] Added Docs for all relevant versions

**Questions/Suggestions**

- Maybe the global config could be moved to `lighthouse.security.paginate_max_count` as it's also related to security (DDOS)
- I'm sure it's possible to simplify tests and implementation by using helpers or refactoring by extracting some parts into methods.
Please let me know what I can improve if you have any suggestions. Thanks :)

### CODE REVIEW

**Comment by spawnia** [comment]
@jbbr this looks pretty good, nice work! If you can add a small example to the docs for the paginate directive, i think this can be merged.

**Comment by lucasmichot** [comment]
@jbbr, what is the difference with `defaultCount`?
 https://github.com/nuwave/lighthouse/blob/28409f24c067b575d3ffcb3577a19f51cdb92e12/src/Schema/Directives/Fields/RelationDirective.php#L57

File: src/Schema/Directives/Fields/PaginationManipulator.php [hunk 1]
```diff
@@ -62,14 +62,14 @@ public static function assertValidPaginationType(string $paginationType): string
      * @param  int|null  $defaultCount
      * @return \Nuwave\Lighthouse\Schema\AST\DocumentAST
```
Comment (lucasmichot) [inline src/Schema/Directives/Fields/PaginationManipulator.php]: Missing docblock param

File: src/Schema/Directives/Fields/PaginationManipulator.php [hunk 2]
```diff
@@ -82,7 +82,7 @@ public static function transformToPaginatedField(string $paginationType, FieldDe
...
```
Comment (lucasmichot) [inline src/Schema/Directives/Fields/PaginationManipulator.php]: same

File: src/Schema/Directives/Fields/PaginationManipulator.php [hunk 3]
```diff
@@ -134,7 +135,7 @@ public static function registerConnection(FieldDefinitionNode $fieldDefinition,
...
```
Comment (lucasmichot) [inline src/Schema/Directives/Fields/PaginationManipulator.php]: same

File: src/Schema/Directives/Fields/PaginateDirective.php [hunk 4]
```diff
@@ -102,6 +110,13 @@ protected function connectionTypeResolver(FieldValue $value): FieldValue
         return $value->setResolver(
             function ($root, array $args): LengthAwarePaginator {
                 $first = $args['first'];
+
```
Comment (lucasmichot) [inline src/Schema/Directives/Fields/PaginateDirective.php]: seems like duplicate code?

File: src/Execution/DataLoader/RelationBatchLoader.php [hunk 5]
```diff
@@ -56,13 +64,14 @@ class RelationBatchLoader extends BatchLoader
      * @param  string|null  $paginationType
      * @return void
```
Comment (lucasmichot) [inline src/Execution/DataLoader/RelationBatchLoader.php]: missing param docblock

**Comment by jbbr** [comment]
@lucasmichot thanks for the review!

> @jbbr, what is the difference with `defaultCount`?
I'm not sure I understand the question. Do you mean in general?

`maxCount` => maximum allowed value for arg `count`, which can be specified by the end-user when querying the paginated field
`defaultCount` => default value for `count`, if none is specified by the end-user

I'm planning to add some documentation and solve your findings this weekend. Sorry for the delay.

**Comment by lucasmichot** [comment]
Thanks for the infos @jbbr 👍 
Also you might want to rebase your branch onto master, so that your tests pass

**Comment by alberthaff** [comment]
Really cool!

We should probably also add a new field in `paginatorInfo` called `maxCount` which either returns the local (if defined) or global maximum.

File: src/Schema/Directives/Fields/PaginateDirective.php (hunk 4 above)
Comment (spawnia) [reply src/Schema/Directives/Fields/PaginateDirective.php]: Deduped

**Comment by spawnia** [comment]
@jbbr just merged the latest changes from master in and refactored a bit. Again, great work done.

Autogenerating the descriptions was a really nice touch.

**Comment by chrissm79** [comment]
Great job on this one @jbbr, thank you!

**Comment by jbbr** [comment]
Thank you for finishing and merging this - and your hard work on this package in general, looking forward to 3.0 :)

**Comment by ZaidBarghouthi** [comment]
Guys! 

So I have `maxCount = 5`
```graphql
allUsers: [User!]! @paginate(maxCount: 5)
```

but when I query with `count: 0` is ignores it.

```graphql
{
  allUsers(count: 0) {
    data {
     id
     name
    }    
  }
}
```

Resutls:

```graphql
{
  "data": {
    "allUsers": {
      "data": [
        {
          "id": "1"
        },
        {
          "id": "2"
        },
        {
          "id": "3"
        },
        {
          "id": "4"
        },
        {
          "id": "5"
        },
        {
          "id": "6"
        },
        {
          "id": "7"
        },
        {
          "id": "8"
        },
        {
          "id": "9"
        },
        {
          "id": "10"
        },
        {
          "id": "11"
        },
        {
          "id": "12"
        },
        {
          "id": "13"
        },
        {
          "id": "14"
        }
      ]
    }
  }
}
```

I guess somewhere there is `if($maxCount)...` causing this, when maxCount is zero it behaves like there in no maxCount!

**Pull request merged**